traitlets/traitlets.py

```diff
@@ -1206,8 +1222,15 @@ def add_traits(self, **traits):
         """Dynamically add trait attributes to the HasTraits instance."""
         self.__class__ = type(self.__class__.__name__, (self.__class__,),
                               traits)
-        for trait in traits.values():
+        
+        observers = [memb[1] for memb in getmembers(self.__class__)
+                     if isinstance(memb[1], ObserveHandler)]
+
+        for name, trait in traits.items():
             trait.instance_init(self)
+            for o in observers:
+                if trait.name in self.traits(**o.tags):
+                    self.observe(o, trait.name, type=o.type)
```

Comment:
This is a good idea but it will only work with observeHandlers and not generic observers registered with the observe method.